File: CTNotificationService/CTNotificationServiceExtension.m

#import "CTNotificationServiceExtension.h"

static NSString * const kMediaUrlKey = @"ct_mediaUrl";
static NSString * const kMediaTypeKey = @"ct_mediaType";

static NSString * const kImage = @"image";
static NSString * const kVideo = @"video";
static NSString * const kAudio = @"audio";
static NSString * const kImageJpeg = @"image/jpeg";
static NSString * const kImagePng = @"image/png";
static NSString * const kImageGif = @"image/gif";

@interface CTNotificationServiceExtension()

@property (nonatomic, strong) void (^contentHandler)(UNNotificationContent *contentToDeliver);
@property (nonatomic, strong) UNMutableNotificationContent *bestAttemptContent;

@end

@implementation CTNotificationServiceExtension

- (void)didReceiveNotificationRequest:(UNNotificationRequest *)request withContentHandler:(void (^)(UNNotificationContent * _Nonnull))contentHandler {
    
    self.contentHandler = contentHandler;
    self.bestAttemptContent = [request.content mutableCopy];
    
    NSDictionary *userInfo = request.content.userInfo;
    if (userInfo == nil) {
        self.contentHandler(self.bestAttemptContent);
        return;
    }
    
    NSString *mediaUrlKey = self.mediaUrlKey ? self.mediaUrlKey : kMediaUrlKey;
    NSString *mediaTypeKey = self.mediaTypeKey ? self.mediaTypeKey : kMediaTypeKey;
    
    NSString *mediaUrl = userInfo[mediaUrlKey];
    NSString *mediaType = userInfo[mediaTypeKey];
    
    if (mediaUrl == nil || mediaType == nil) {
#ifdef DEBUG
        if (mediaUrl == nil) {
            NSLog(@"unable to add attachment: %@ is nil", mediaUrlKey);
        }
        
        if (mediaType == nil) {
            NSLog(@"unable to add attachment: %@ is nil", mediaTypeKey);
        }
        
#endif
        self.contentHandler(self.bestAttemptContent);
        return;
    }
    
    // load the attachment
    [self loadAttachmentForUrlString:mediaUrl
                            withType:mediaType
                   completionHandler:^(UNNotificationAttachment *attachment) {
        if (attachment) {
            self.bestAttemptContent.attachments = [NSArray arrayWithObject:attachment];
        }
        self.contentHandler(self.bestAttemptContent);
    }];
    
}

- (void)serviceExtensionTimeWillExpire {
    self.contentHandler(self.bestAttemptContent);
}

- (NSString *)fileExtensionForMediaType:(NSString *)mediaType mimeType:(NSString *)mimeType {
    NSString *ext;
    
    if ([mediaType isEqualToString:kImage]) {
        ext = @"jpg";
    } else if ([mediaType isEqualToString:kVideo]) {
        ext = @"mp4";
    } else if ([mediaType isEqualToString:kAudio]) {
        ext = @"mp3";
    } else {
        // If mediaType is none, check for mimeType of url.
        if ([mimeType isEqualToString:kImageJpeg]) {
            ext = @"jpeg";
        } else if ([mimeType isEqualToString:kImagePng]) {
            ext = @"png";
        } else if ([mimeType isEqualToString:kImageGif]) {
            ext = @"gif";
        } else {
            ext = @"";
        }
    }
    
    return [@"." stringByAppendingString:ext];
}

- (void)loadAttachmentForUrlString:(NSString *)urlString withType:(NSString *)mediaType
                 completionHandler:(void(^)(UNNotificationAttachment *))completionHandler  {
    
    __block UNNotificationAttachment *attachment = nil;
    NSURL *attachmentURL = [NSURL URLWithString:urlString];
    
    NSURLSession *session = [NSURLSession sessionWithConfiguration:[NSURLSessionConfiguration defaultSessionConfiguration]];
    [[session downloadTaskWithURL:attachmentURL
                completionHandler:^(NSURL *temporaryFileLocation, NSURLResponse *response, NSError *error) {
        if (error != nil) {
#ifdef DEBUG
            NSLog(@"unable to add attachment: %@", error.localizedDescription);
#endif
        } else {
            NSString *fileExt = [self fileExtensionForMediaType:mediaType mimeType:response.MIMEType];
            NSFileManager *fileManager = [NSFileManager defaultManager];
            NSURL *localURL = [NSURL fileURLWithPath:[temporaryFileLocation.path stringByAppendingString:fileExt]];
            [fileManager moveItemAtURL:temporaryFileLocation toURL:localURL error:&error];
            
            NSError *attachmentError = nil;
            attachment = [UNNotificationAttachment attachmentWithIdentifier:@"" URL:localURL options:nil error:&attachmentError];
            if (attachmentError) {
#ifdef DEBUG
                NSLog(@"unable to add attchment: %@", attachmentError.localizedDescription);
#endif
            }
        }
        completionHandler(attachment);
    }] resume];
}


@end

File: CTNotificationService/CTNotificationServiceExtension.swift
import UserNotifications

open class CTNotificationServiceExtension: UNNotificationServiceExtension {

    var contentHandler: ((UNNotificationContent) -> Void)?
    var bestAttemptContent: UNMutableNotificationContent?

    open override func didReceive(_ request: UNNotificationRequest, withContentHandler contentHandler: @escaping (UNNotificationContent) -> Void) {

          self.contentHandler = contentHandler
          bestAttemptContent = (request.content.mutableCopy() as? UNMutableNotificationContent)
          
          // MARK: - CT Rich Push
          if let bestAttemptContent = bestAttemptContent {
//              bestAttemptContent.title = "\(bestAttemptContent.title) [modified]"
              let userInfo = request.content.userInfo;

              // If there is no URL in the payload than
              // the code will still show the push notification.
              if userInfo["ct_mediaUrl"] == nil {
                  contentHandler(bestAttemptContent);
                  return;
              }
              
              let mediaType = userInfo["ct_mediaType"] as? String
              let mediaUrl = userInfo["ct_mediaUrl"] as? String
              
              if (mediaUrl != nil) {
                  let mediaUrl = URL(string: mediaUrl!)
                  let CTSession = URLSession(configuration: .default)
                  CTSession.downloadTask(with: mediaUrl!, completionHandler: { temporaryLocation, response, error in
                      if let err = error {
                          print("Error with downloading rich push: \(String(describing: err.localizedDescription))")
                          contentHandler(bestAttemptContent);
                          return;
                      }
                      
                      let fileType = self.fileExtensionForMediaType(fileType: (response?.mimeType)!)
                      let fileName = temporaryLocation?.lastPathComponent.appending(fileType)
                      
                      let temporaryDirectory = URL(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent(fileName!)
                      
                      do {
                          try FileManager.default.moveItem(at: temporaryLocation!, to: temporaryDirectory)
                          let attachment = try UNNotificationAttachment(identifier: "", url: temporaryDirectory, options: nil)
                          
                          bestAttemptContent.attachments = [attachment];
                          contentHandler(bestAttemptContent);
                          // The file should be removed automatically from temp
                          // Delete it manually if it is not
                          if FileManager.default.fileExists(atPath: temporaryDirectory.path) {
                              try FileManager.default.removeItem(at: temporaryDirectory)
                          }
                      } catch {
                          print("Error with the rich push attachment: \(error)")
                          contentHandler(bestAttemptContent);
                          return;
                      }
                  }).resume()
                  
              }
          }
      }
      
    open override func serviceExtensionTimeWillExpire() {
          // Called just before the extension will be terminated by the system.
          // Use this as an opportunity to deliver your "best attempt" at modified content, otherwise the original push payload will be used.
          if let contentHandler = contentHandler, let bestAttemptContent =  bestAttemptContent {
              contentHandler(bestAttemptContent)
          }

}

    func fileExtensionForMediaType(fileType: String) -> String {
        // Determines the file type of the attachment to append to URL.
        if fileType == "image/jpeg" {
            return ".jpg";
        }
        if fileType == "image/gif" {
            return ".gif";
        }
        if fileType == "video/mp4" {
            return ".mp4";
        }
        if fileType == "audio/mpeg" {
            return ".mpeg";
        }
        if fileType == "image/png" {
            return ".png";
        } else {
            return ".tmp";
        }
    }

}
